blazar/plugins/devices/device_plugin.py

```diff
@@ -710,10 +736,210 @@ def _allocations_to_remove(self, dates_before, dates_after, max_devices,
         return allocs_to_remove
 
     def _filter_devices_by_properties(self, resource_properties):
-        filter = []
         if resource_properties:
             filter += plugins_utils.convert_requirements(resource_properties)
         if filter:
             return db_api.device_get_all_by_queries(filter)
         else:
             return db_api.device_list()
+
+    def heal_reservations(self, failed_resources, interval_begin,
+                          interval_end):
+        """Heal reservations which suffer from resource failures.
+
+        :param failed_resources: a list of failed devices.
+        :param interval_begin: start date of the period to heal.
+        :param interval_end: end date of the period to heal.
+        :return: a dictionary of {reservation id: flags to update}
+                 e.g. {'de27786d-bd96-46bb-8363-19c13b2c6657':
+                       {'missing_resources': True}}
+        """
+        reservation_flags = {}
+
+        device_ids = [h['id'] for h in failed_resources]
+        reservations = db_utils.get_reservations_by_device_ids(device_ids,
+                                                               interval_begin,
+                                                               interval_end)
+
+        for reservation in reservations:
+            if reservation['resource_type'] != plugin.RESOURCE_TYPE:
+                continue
+
+            for allocation in [alloc for alloc
+                               in reservation['device_allocations']
+                               if alloc['device_id'] in device_ids]:
+                if self._reallocate(allocation):
+                    if reservation['status'] == status.reservation.ACTIVE:
+                        if reservation['id'] not in reservation_flags:
```

Comment:
I think reservation['id'] could be pulled out to a variable given how often it is used